src/compiler/crystal/codegen/debug.cr

```diff
@@ -110,40 +132,181 @@ module Crystal
+      ivars = type.entries
+      element_types = [] of LibLLVMExt::Metadata
+      struct_type = llvm_struct_type(type)
+
+      tmp_debug_type = di_builder.create_replaceable_composite_type(nil, orig_type.to_s, nil, 1, llvm_context)
+      debug_type_cache[orig_type] = tmp_debug_type
+
+      ivars.each_with_index do |ivar, idx|
+        next if (ivar_type = ivar.type).is_a?(NilType)
+        if ivar_debug_type = get_debug_type(ivar_type)
+          offset = @program.target_machine.data_layout.offset_of_element(struct_type, idx &+ (type.struct? ? 0 : 1))
+          size = @program.target_machine.data_layout.size_in_bits(llvm_embedded_type(ivar_type))
+
+          # FIXME structs like LibC::PthreadMutexT generate huge offset values
+          next if offset > UInt64::MAX // 8u64
+
+          member = di_builder.create_member_type(nil, ivar.name, nil, 1, size, size, 8u64 * offset, LLVM::DIFlags::Zero, ivar_debug_type)
+          element_types << member
+        end
+      end
+
+      size = @program.target_machine.data_layout.size_in_bits(struct_type)
+      debug_type = di_builder.create_struct_type(nil, orig_type.to_s, nil, 1, size, size, LLVM::DIFlags::Zero, nil, di_builder.get_or_create_type_array(element_types))
+      unless type.struct?
+        debug_type = di_builder.create_pointer_type(debug_type, 8u64 * llvm_typer.pointer_size, 8u64 * llvm_typer.pointer_size, orig_type.to_s)
+      end
+      di_builder.replace_temporary(tmp_debug_type, debug_type)
+      debug_type
+    end
+
+    # This is a sinkhole for debug types that most likely does not need to be implemented
+    def create_debug_type(type : NonGenericModuleType | GenericClassInstanceMetaclassType | MetaclassType | NilableProcType | VirtualMetaclassType, orig_type : Type)
+    end
+
+    def create_debug_type(type, orig_type : Type)
+      # "Unsupported type for debugging: #{type} (#{type.class}), orig_type=#{orig_type.to_s}" }
     end
 
     def declare_parameter(arg_name, arg_type, arg_no, alloca, location)
+      return alloca unless @debug.variables?
+
       declare_local(arg_type, alloca, location) do |scope, file, line_number, debug_type|
         di_builder.create_parameter_variable scope, arg_name, arg_no, file, line_number, debug_type
       end
     end
 
-    def declare_variable(var_name, var_type, alloca, location)
-      declare_local(var_type, alloca, location) do |scope, file, line_number, debug_type|
+    def declare_variable(var_name, var_type, alloca, location, basic_block : LLVM::BasicBlock? = nil)
+      return false unless @debug.variables?
+      declare_local(var_type, alloca, location, basic_block) do |scope, file, line_number, debug_type|
         di_builder.create_auto_variable scope, var_name, file, line_number, debug_type, align_of(var_type)
       end
     end
 
+    def dump_metadata(md : LibLLVMExt::Metadata?)
+      return "nil" unless md
+      LLVM::Value.new(LibLLVM.metadata_as_value(llvm_context, md))
+    end
+
```

Comment:
This seems to be unused...?

Reply:
I used to use it for debugging when debugging support was not implemented yet. It was showing very useful information. I can move it into LLVM module though as it can be useful for programmers who will use LLVM for their own needs.

Reply:
Removed.